internal/encryption/targets.go

```diff
@@ -9,121 +9,75 @@ import (
-// The targetName parameter is used for error messages only.
-func (e *targetBuilder) build(target *config.TargetConfig, targetName string) (methods []method.Method, diags hcl.Diagnostics) {
+			mRoot, okRoot := traversal[0].(hcl.TraverseAttr)
+			mType, okType := traversal[1].(hcl.TraverseAttr)
+			mName, okName := traversal[2].(hcl.TraverseAttr)
 
-	// gohcl has some weirdness around attributes that are not provided, but are hcl.Expressions
-	// They will set the attribute field to a static null expression
-	// https://github.com/hashicorp/hcl/blob/main/gohcl/decode.go#L112-L118
+			if !okRoot || mRoot.Name != "method" || !okType || !okName {
+				diags = diags.Append(&hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Invalid encryption method identifier",
+					Detail:   "Expected method of form method.<type>.<name>",
+					Subject:  target.Method.Range().Ptr(),
+				})
+			}
 
-	// Descriptor referenced by this target
-	var methodIdent *string
-	decodeDiags := gohcl.DecodeExpression(target.Method, e.ctx, &methodIdent)
-	diags = append(diags, decodeDiags...)
+			foundMethod := false
+			for _, method := range cfg.MethodConfigs {
+				if method.Type == mType.Name && method.Name == mName.Name {
+					foundMethod = true
+					methods = append(methods, method)
+					break
+				}
+			}
 
-	// Only attempt to fetch the method if the decoding was successful
-	if !decodeDiags.HasErrors() {
-		if methodIdent != nil {
-			if method, ok := e.methods[method.Addr(*methodIdent)]; ok {
-				methods = append(methods, method)
-			} else {
+			if !foundMethod {
 				// We can't continue if the method is not found
 				diags = append(diags, &hcl.Diagnostic{
 					Severity: hcl.DiagError,
 					Summary:  "Undefined encryption method",
-					Detail:   fmt.Sprintf("Can not find %q for %q", *methodIdent, targetName),
+					Detail:   fmt.Sprintf("Can not find \"%s.%s.%s\" for %s", mRoot.Name, mType.Name, mName.Name, targetName),
 					Subject:  target.Method.Range().Ptr(),
 				})
 			}
-		} else {
-			diags = append(diags, &hcl.Diagnostic{
-				Severity: hcl.DiagError,
-				Summary:  "Missing encryption method",
-				Detail:   fmt.Sprintf("undefined or null method used for %q", targetName),
-				Subject:  target.Method.Range().Ptr(),
-			})
 		}
+
+		// Attempt to fetch the fallback method if it's been configured
+		targetName += ".fallback"
+		target = target.Fallback
 	}
 
-	// Attempt to fetch the fallback method if it's been configured
-	if target.Fallback != nil {
-		fallback, fallbackDiags := e.build(target.Fallback, targetName+".fallback")
-		diags = append(diags, fallbackDiags...)
-		methods = append(methods, fallback...)
+	if enforced {
+		for _, m := range methods {
+			if unencrypted.IsConfig(m) {
+				return nil, append(diags, &hcl.Diagnostic{
+					Severity: hcl.DiagError,
+					Summary:  "Unencrypted method is forbidden",
+					Detail:   "Unable to use `unencrypted` method since the `enforced` flag is used.",
```

Comment:
OpenTofu error messages don't typically use backticks like this. I would suggest either leaving them unquoted or using normal double quotes for consistency. I might also say that the enforced flag is _true_, rather than just that it is _used_, to be clearer what exactly we're talking about.